File: 3960-frequency-balance-subarray/3960-frequency-balance-subarray.py
from collections import Counter
from typing import List

class Solution:
    def getLength(self, nums: List[int]) -> int:
        n = len(nums)
        ml=1
        for i in range(n):
            ct=Counter()
            fm=Counter()
            for j in range(i,n):
                val = nums[j]
                if ct[val]>0:
                    oldf=ct[val]
                    fm[oldf]-=1
                    if fm[oldf]==0:
                        del fm[oldf]
                ct[val]+=1
                newf=ct[val]
                fm[newf]+=1
                cl=j-i+1
                if cl==1:
                    continue
                if len(fm)==1:
                    freq=next(iter(fm))
                    if fm[freq]==1:
                        ml=max(ml,cl)
                elif len(fm)== 2:
                    sf=sorted(fm.keys())
                    minf,maxf=sf[0],sf[1]
                    if maxf==2*minf:
                        ml=max(ml,cl)
            
                    
        return ml

File: 3960-frequency-balance-subarray/3960-frequency-balance-subarray.cpp
class Solution {
public:
    int getLength(vector<int>& a) {
        int n=a.size(),r=1;
        for(int i=0;i<n;i++){
            unordered_map<int,int> c,f;
            for(int j=i;j<n;j++){
                int x=a[j],p=c[x]++;
                p&&(!--f[p]&&f.erase(p),0);
                f[p+1]++;
                int l=j-i+1,ok=l<2||c.size()<2;
                if(!ok&&f.size()==2){
                    auto u=f.begin(),v=next(u);
                    int x=u->first,y=v->first;
                    ok=(x==(y<<1))||(y==(x<<1));
                }
                ok&&(r=max(r,l));
            }
        }
        return r;
    }
};